src/relay/backend/contrib/tensorrt/trt_builder.cc

```diff
@@ -0,0 +1,392 @@
+        {"divide", new ElementWiseBinaryOpConverter()},
+        {"power", new ElementWiseBinaryOpConverter()},
+        {"nn.max_pool2d", new PoolingOpConverter()},
+        {"nn.avg_pool2d", new PoolingOpConverter()},
+        {"nn.global_max_pool2d", new GlobalPoolingOpConverter()},
+        {"nn.global_avg_pool2d", new GlobalPoolingOpConverter()},
+        {"exp", new UnaryOpConverter()},
+        {"log", new UnaryOpConverter()},
+        {"sqrt", new UnaryOpConverter()},
+        {"abs", new UnaryOpConverter()},
+        {"negative", new UnaryOpConverter()},
+        {"nn.batch_flatten", new BatchFlattenOpConverter()},
+        {"expand_dims", new ExpandDimsOpConverter()},
+        {"squeeze", new SqueezeOpConverter()},
+        {"concatenate", new ConcatOpConverter()},
+        {"nn.conv2d_transpose", new Conv2DTransposeOpConverter()},
+        {"transpose", new TransposeOpConverter()},
+        {"reshape", new ReshapeOpConverter()},
+        {"nn.pad", new PadOpConverter()},
+        {"sum", new ReduceOpConverter()},
+        {"prod", new ReduceOpConverter()},
+        {"max", new ReduceOpConverter()},
+        {"min", new ReduceOpConverter()},
+        {"mean", new ReduceOpConverter()},
+        {"contrib.adaptive_max_pool2d", new AdaptivePoolingOpConverter()},
+        {"contrib.adaptive_avg_pool2d", new AdaptivePoolingOpConverter()},
+#if TRT_VERSION_GE(5, 1, 5)
+        {"clip", new ActivationOpConverter()},
+        {"nn.leaky_relu", new ActivationOpConverter()},
+        {"sin", new UnaryOpConverter()},
+        {"cos", new UnaryOpConverter()},
+        {"atan", new UnaryOpConverter()},
+        {"ceil", new UnaryOpConverter()},
+        {"floor", new UnaryOpConverter()},
+        {"strided_slice", new StridedSliceOpConverter()},
+#endif
+};
+
+TrtBuilder::TrtBuilder(const std::vector<DLTensor*>& args)
+    : execution_args_(args) {
+  // Create TRT builder and network.
+  static TensorRTLogger trt_logger;
+  builder_ = nvinfer1::createInferBuilder(trt_logger);
+  const int batch_size = args[0]->shape[0];
+  builder_->setMaxBatchSize(batch_size);
+  const size_t workspace_size =
+      dmlc::GetEnv("TVM_TENSORRT_MAX_WORKSPACE_SIZE", size_t(1) << 31);
+  builder_->setMaxWorkspaceSize(workspace_size);
+  const bool use_fp16 = dmlc::GetEnv("TVM_TENSORRT_USE_FP16", false);
+  builder_->setFp16Mode(use_fp16);
+  network_ = builder_->createNetwork();
+}
+
+TrtEngineAndContext TrtBuilder::BuildEngine(const Expr& expr) {
+  // Process graph and create INetworkDefinition.
+  VisitExpr(expr);
+  // Mark outputs.
+  auto network_outputs = node_output_map_[expr.operator->()];
+  std::vector<std::string> network_output_names;
+  for (int i = 0; i < network_outputs.size(); ++i) {
+    CHECK(network_outputs[i].type == kTensor);
+    auto out_tensor = network_outputs[i].tensor;
+    std::string output_name = "tensorrt_output" + std::to_string(i);
+    out_tensor->setName(output_name.c_str());
+    network_output_names.push_back(output_name);
```

Comment:
I am not quite sure if we want to directly interact with trt tensor here. We provided an minimal example runtime (https://github.com/apache/incubator-tvm/pull/4280/) for JSON and codegen for DNNL(https://github.com/apache/incubator-tvm/pull/4482) 

Maybe we can make it closer to what we have so that the maintenance could be easier in the future

Reply:
I have updated the code in a separate branch to use the external codegen/runtime. Here is the branch: https://github.com/trevor-m/tvm/tree/trt-new
and commit to use external codegen/runtime: https://github.com/trevor-m/tvm/commit/595cad29cbeb01454e9a18a19d33da5dda96e4cf

However, neo-ai/tvm doesn't have the necessary code yet. I will update the PR once neo-ai/tvm is synced to upstream.

This eliminates all changes to:
src/relay/backend/graph_runtime_codegen.cc 
src/runtime/graph/graph_runtime.cc 
src/runtime/graph/graph_runtime.h

Reply:
@zxy844288792 When can we sync to upstream master?

Reply:
It is a rotation to sync to upstream and we should ask Lokesh about the plans.